server/modules/express-middleware/deprecateMiddleware.js

```diff
@@ -0,0 +1,26 @@
+/* Copyright (c) Trainline Limited, 2016-2017. All rights reserved. See LICENSE.txt in the project root for license information. */
+
+/**
+ * Express middleware that adds deprecated=true to a request and sets the Warning HTTP header
+ * on the response if it is a request to a deprecated route.
+ */
+
+'use strict';
+
+function create(fn) {
+  return function deprecateMiddleware(req, res, next) {
+    try {
+      let warning = fn(req);
+      if (warning) {
+        let now = new Date().toUTCString();
+        res.locals.deprecated = true;
+        res.append('Warning', `299 - Deprecated: ${warning} "${now}"`);
```

Comment:
I agree that would be nice but the most important requirement is not to break the existing pre v1 API. I considered it and decided against for these reasons:
* Only GET and HEAD methods can't be redirected but the API relies heavily on POST, PUT etc.
* There is often not a corresponding v1 API URL to which to the user agent could be redirected.
* User agents are within their rights to strip auth headers before following redirect. In particular [`System.Net.Http.HttpClient`](https://www.google.co.uk/url?sa=t&rct=j&q=&esrc=s&source=web&cd=1&cad=rja&uact=8&ved=0ahUKEwj41suS5ZTSAhWB2BoKHRgWANsQFggaMAA&url=https%3A%2F%2Fmsdn.microsoft.com%2Fen-us%2Flibrary%2Fsystem.net.http.httpclient(v%3Dvs.118).aspx&usg=AFQjCNHCCYB1s1G4im9Vh_sTYvSGwHhHZw&sig2=TT2FAt1ydnFphATQePBR-g&bvm=bv.147134024,d.d2s) will do this.